Composer/packages/server/src/models/asset/assetManager.ts

```diff
@@ -92,58 +97,86 @@ export class AssetManager {
     return ref;
   }
 
-  private async getRemoteTemplate(template: BotTemplate, destinationPath: string) {
-    // install package
-    if (template.package) {
-      const { stderr: initErr } = await execAsync(`dotnet new -i ${template.package.packageName}`);
-      if (initErr) {
-        throw new Error(initErr);
+  public async copyRemoteProjectTemplateToV2(
+    templateId: string,
+    templateVersion: string,
+    projectName: string,
+    ref: LocationRef,
+    user?: UserIdentity
+  ): Promise<LocationRef> {
+    try {
+      // user storage maybe diff from template storage
+      const dstStorage = StorageService.getStorageClient(ref.storageId, user);
+      const dstDir = Path.resolve(ref.path);
+      if (await dstStorage.exists(dstDir)) {
+        log('Failed copying template to %s', dstDir);
+        throw new Error('already have this folder, please give another name');
       }
-      const { stderr: initErr2 } = await execAsync(`dotnet new ${template.id}`, {
-        cwd: destinationPath,
-      });
-      if (initErr2) {
-        throw new Error(initErr2);
+
+      mkdirSync(dstDir, { recursive: true });
+
+      // find selected template
+      const npmPackageName = templateId;
+      const generatorName = npmPackageName.toLowerCase().replace('generator-', '');
+
+      const remoteTemplateAvailable = await this.installRemoteTemplate(generatorName, npmPackageName, templateVersion);
+
+      if (remoteTemplateAvailable) {
+        await this.instantiateRemoteTemplate(generatorName, dstDir, projectName);
       }
-    } else {
-      throw new Error('selected template has no local or external address');
+
+      ref.path = `${ref.path}/${projectName}`;
+
+      return ref;
+    } catch (err) {
+      throw new Error(`error hit when instantiating remote template: ${err?.message}`);
+    }
+  }
+
+  private async installRemoteTemplate(
+    generatorName: string,
+    npmPackageName: string,
+    templateVersion: string
+  ): Promise<boolean> {
+    this.yeomanEnv.cwd = templateGeneratorPath;
+    try {
+      log('Installing generator', npmPackageName);
+      templateVersion = templateVersion ? templateVersion : '*';
+      await this.yeomanEnv.installLocalGenerators({ [npmPackageName]: templateVersion });
+
+      log('Looking up local packages');
+      await this.yeomanEnv.lookupLocalPackages();
+      return true;
+    } catch (err) {
+      throw new Error(`error hit when installing remote template: ${err?.message}`);
     }
   }
 
+  private async instantiateRemoteTemplate(
+    generatorName: string,
+    dstDir: string,
+    projectName: string
+  ): Promise<boolean> {
+    this.yeomanEnv.cwd = dstDir;
+
+    log('About to instantiate a template!', dstDir, generatorName, projectName);
```

Comment:
does awaiting this work?


normally when there is a callback, it doesn't work as a promise.

We should double check on this - otherwise, wrap this in a Promise and resolve inside the callback.

Reply:
Just ran a test, awaiting it does work :D. It runs the callback prior to continuing the function and returning 'true'